### PR TITLE
feat: add tooltip component

### PR DESCRIPTION
**Pull request by scomea**
# Description

This pull request adds a tooltip component.

noteworthy bits:
- it uses anchored region for placement
- tooltips render with fixed placement so they break out of parent containers and can easily appear on top of everything with the right z-index.  Down side they don't scroll with their parent container.  This seems desirable as a default behavior to me. Is it? Do we (initially) need other options like being able to associate tooltips with a different viewport so they could move with scrolled content, be clipped by parent container, etc...?  
- current default placement logic is above or below the anchor depending on available space.  Authors can specify top/right/bottom/left. Is this adequate for the initial version?  Anchored region supports much finer positioning control so at a minimum we should ensure that we don't need to make breaking changes to take advantage later.
- tooltips are super thin until they are shown, before then they just render the outer template to the dom and run a minimal amount of code.  This should be excellent for loading perf on pages that have many of these, but that does mean items in the tooltip slot are not rendered to the dom before then.  Does this need to be controllable?
 

## Motivation & context
We want a tooltip component

## Issue type checklist
- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

**Adding or modifying component(s) in `@microsoft/fast-components` checklist*
- [x] I have added a new component
- [ ] I have modified an existing component
- [x] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [x] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## Process & policy checklist
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

**Comment by EisenbergEffect** [comment]
My thoughts on the questions:

* I don't think tooltips need to scroll. I did a quick test and from what I can tell the native tooltips don't scroll.
* Current placement behavior and options seem good to me.
* I don't think we need to render the tooltip slot content when it's now showing. I wouldn't make this controllable in our first version. If we get significant customer request for that, we could add it later. Not sure why we would really need that though.

File: packages/web-components/fast-foundation/src/tooltip/tooltip.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: Can't rely on being able to find it in the document. Might need to use `this.getRootNode()`.

Comment (scomea) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: Even when wrapped in a "$fastController.isConnected" check?

Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: No, because if your component is inside of another component's shadow dom, a selector executed against the document will not be able to fin it. You need to get the root node (which might be the document...but could be another shadow root) and then run the query selector on that.

Comment (scomea) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: So whenever we want to call something on document from code in a web component we should recursively call getRootNode until the return type is HTMLDocument?

Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: You don't need to recursively call it. Calling it once will get you the nearest shadow root that your element is in, or the document if it's not in a shadow DOM.

Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: Maybe I misunderstood the code...are you trying to find all :hover in the same "scope" as the current element? If so, then you would use getRootNode()

Comment (scomea) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: No, I'm actually trying to figure out if the anchor is being hovered at that point in time. It could be anywhere in the document.

Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: In the document or in the DOM scope that the tooltip is being used within?

Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: There can be N doms in a web page. The document is one but so is each shadow DOM. The tooltip could exist directly in the document or also inside of a shadow dom. My assumption is that the tooltip is used in the same dom as its anchor, in which case you would need to use getRootNode(). If the anchor can be anywhere...then we probably need to change the design of this component or at least introduce the restriction that the tooltip and the anchor need to exist in the same DOM.

Comment (scomea) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: Ok, so I can query the anchor element itself to determine hover, which is actually way better than querying the document.  But there are other places where we query the document for elements by id to get the anchor and viewport.  I think that's ok as that's the most basic use case and we present more advanced authors who may be working in a shadow dom with the ability to pass in those elements directly via JS.

Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:68]: I think that sounds fine.

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (chrisdholt) [inline packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:29]: Can we drop the accent fill behavior from line 28, assuming line 29 is overriding it.

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (chrisdholt) [inline packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:40]: I see the 10000, I wonder if we should create a semantic ramp for values that we can invoke. This also may be beneficial as a var with a default value. That would make it easy to manage as the owner of an app - especially if CSS Parts are not supported in some of my primary browsers.

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (chrisdholt) [inline packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:74]: Will these need to localize? Perhaps instead of left/right we should use language similar to the CSS logical properties and map to those? Curious on thoughts from @EisenbergEffect, @nicholasrice, others?

```
:host(.inline-start) {
    margin-inline-start: 4px;
}
```

File: packages/web-components/fast-foundation/src/tooltip/README.md
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/tooltip/README.md:8]: Should this be filled out?

File: packages/web-components/fast-foundation/src/tooltip/tooltip.spec.md
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/tooltip/tooltip.spec.md:33]: Thanks! 😄

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (EisenbergEffect) [reply packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:74]: That makes good sense to me. If that's not enough we also have the new RTL behaviors that Nick wrote. Those are going to be more heavy weight than a standard CSS property though. Just mentioning it for completeness.

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (EisenbergEffect) [reply packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:40]: I think we should have a central way to manage the zindex for modals, tooltips, select popups, etc. For example, what if we have a modal with a select in it and there's a tooltip on something in the select popup? Do we alway treat tooltips as highest, then popups, then modals? Would something simple like that work? Can we just encode that in a simple lookup in foundation that can be used by various components and design systems?

Comment (chrisdholt) [reply packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:40]: Yeah, there are a few ways that we've done this before - I'm happy to open an issue and propose something.

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (chrisdholt) [reply packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:74]: For sure - I like the logical aspect of the naming; we did something similar with the start/end slots.

Comment (scomea) [reply packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:74]: The current configuration works in rtl.  The host "left"/"right" classes are applied based on the actual position of the region relative to the anchor.  If the region is left we want a margin on the right and vice versa.

Comment (scomea) [reply packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:74]: Anchored region supports "start/end" as a choice for horizontal position when authors want position to change based on direction.  I shall add that as a position choice as well as "left/right".

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (chrisdholt) [reply packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:40]: I think it's fine for now - we don't have to block on this. Could you add a note w/ a TODO so I can find it :)

File: packages/web-components/fast-components/src/tooltip/fixtures/base.html
Comment (awentzel) [inline packages/web-components/fast-components/src/tooltip/fixtures/base.html:4]: Our headings should be semantic for those using screen readers.  They seem to be used stylistically when they should be used contextually and semantically.  So H4 really should be H2.  Additionally, should we be attributing with styles?  As documentation, should we be using our own design system just like a production scenario would?

File: packages/web-components/fast-components/src/tooltip/tooltip.stories.ts
Comment (awentzel) [inline packages/web-components/fast-components/src/tooltip/tooltip.stories.ts:7]: I'd prefer to see an explanation of why we are preventing tree-shaking in the comment.

File: packages/web-components/fast-components/src/tooltip/fixtures/base.html
Comment (chrisdholt) [reply packages/web-components/fast-components/src/tooltip/fixtures/base.html:4]: Can we create an issue to do a larger pass here? This is an issue across all controls and not specific to this one, outside that I agree - this is net new and should be correct.

File: packages/web-components/fast-components/src/tooltip/tooltip.stories.ts
Comment (chrisdholt) [reply packages/web-components/fast-components/src/tooltip/tooltip.stories.ts:7]: I think this comment is clear enough for the stories page. We prevent tree shaking here because without the Class defined or used in this it will be tree shaken out.

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (awentzel) [inline packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:39]: Can you link the corresponding `todo` with an GH issue.

**Comment by EisenbergEffect** [comment]
@chrisdholt Did you want to try to get this in for Thursday's release? Anything else not assigned to a release that you want to try to get in?

**Comment by chrisdholt** [comment]
> @chrisdholt Did you want to try to get this in for Thursday's release? Anything else not assigned to a release that you want to try to get in?

Would love this to make it into this release, but *only* if it's ready. Let's not hold for it, but would be great to have it in!

**Comment by scomea** [comment]
Let's get the review moving forward without  rushing.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by marjonlynch** [comment]
I think you want this showing up in fast-component-explorer and in the Docs, I'm not seeing it in either. I'm pretty sure at least you will need to update sidebar.js for docs among other things

File: packages/web-components/fast-foundation/src/tooltip/tooltip.ts
Comment (scomea) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:40]: I think leaving it as undefined is fine.  Not sure about how we should handle the @defaultValue when that is the case tho.

File: packages/web-components/fast-components/src/tooltip/fixtures/base.html
Comment (scomea) [reply packages/web-components/fast-components/src/tooltip/fixtures/base.html]: https://github.com/microsoft/fast/issues/3812

File: packages/web-components/fast-components/src/tooltip/tooltip.styles.ts
Comment (scomea) [reply packages/web-components/fast-components/src/tooltip/tooltip.styles.ts:39]: https://github.com/microsoft/fast/issues/3813

**Comment by scomea** [comment]
Should show up in the component explorer now

**Review by marjonlynch — APPROVED** [review]
Approving with 2 suggestions
1. Update the component explorer example to have a component with the tooltip anchored instead of just a tooltip
2. Write a spec based on Chris's PR that just went in adding all the specs

File: packages/web-components/fast-foundation/src/tooltip/tooltip.ts
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/tooltip/tooltip.ts:231]: You don't need the constructor if you're only calling `super()`

File: packages/web-components/fast-foundation/src/tooltip/tooltip.ts
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/tooltip/tooltip.ts:255]: Is there a reason this work isn't being done in the template where it is batched?

File: packages/web-components/fast-foundation/src/tooltip/tooltip.ts
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/tooltip/tooltip.ts:460]: There is a utility in fast-foundation for this now; `getDirection`

**Review by nicholasrice — APPROVED** [review]
Looks good! Left a few notes but signed off

File: packages/web-components/fast-foundation/src/tooltip/README.md
Comment (chrisdholt) [reply packages/web-components/fast-foundation/src/tooltip/README.md:8]: @scomea - *did this get filled out*?

File: packages/web-components/fast-foundation/src/tooltip/tooltip.ts
Comment (scomea) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts]: switched to that, thx!

File: packages/web-components/fast-foundation/src/tooltip/tooltip.ts
Comment (scomea) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts]: There used to be some code there, removedd.

File: packages/web-components/fast-foundation/src/tooltip/tooltip.ts
Comment (scomea) [reply packages/web-components/fast-foundation/src/tooltip/tooltip.ts:252]: I like that this approach for core functionality does not depend on the template side implementation which can be swapped out.  Should revisit if we suspect a perf cost though.

**Review by chrisdholt — APPROVED** [review]

**Comment by EisenbergEffect** [comment]
@scomea Can you do a follow on PR to make sure the readme matches the other components and then add this component into the list of components in `sidebars.js` in the site? Also, if we want to ship anchored region for external use we'd need to do the same with that as well as add it from the exclude in the `prepare-docs.js` script.

**Pull request merged**
